File: backend/test_api.py
#!/usr/bin/env python
import json
import sys
import subprocess
from pathlib import Path
from datetime import datetime
import requests
import threading

API_BASE_URL = "http://localhost:8000"

def test_api_endpoint(endpoint, method="GET", data=None, expected_status=200):
    """测试API端点"""
    url = f"{API_BASE_URL}{endpoint}"
    print(f"\n测试 {method} {url}")
    
    try:
        if method == "GET":
            response = requests.get(url)
        elif method == "POST":
            response = requests.post(url, json=data)
        else:
            print(f"不支持的HTTP方法: {method}")
            return False
        
        status_code = response.status_code
        print(f"状态码: {status_code}")
        
        try:
            response_json = response.json()
            print(f"响应内容: {json.dumps(response_json, ensure_ascii=False, indent=2)}")
        except:
            print(f"响应内容: {response.text}")
        
        if status_code != expected_status:
            print(f"警告: 状态码 {status_code} 与预期的 {expected_status} 不符")
            return False
        
        return True
    except Exception as e:
        print(f"请求失败: {e}")
        return False

def test_all_endpoints():
    """测试所有API端点"""
    successes = 0
    failures = 0
    
    # 测试API状态
    if test_api_endpoint("/api/status"):
        successes += 1
    else:
        failures += 1
    
    # 测试学习内容API
    if test_api_endpoint("/api/learning", expected_status=500):
        successes += 1
    else:
        failures += 1
    
    # 测试新闻API
    if test_api_endpoint("/api/news", expected_status=500):
        successes += 1
    else:
        failures += 1
    
    # 测试问题API
    if test_api_endpoint("/api/questions", expected_status=500):
        successes += 1
    else:
        failures += 1
    
    # 测试搜索API
    search_data = {"query": "基金", "categories": []}
    if test_api_endpoint("/api/search", method="POST", data=search_data, expected_status=500):
        successes += 1
    else:
        failures += 1
    
    # 测试聊天API
    chat_data = {"message": "什么是股票?"}
    if test_api_endpoint("/api/chat", method="POST", data=chat_data):
        successes += 1
    else:
        failures += 1
    
    # 打印测试摘要
    print(f"\n===== 测试摘要 =====")
    print(f"成功: {successes}")
    print(f"失败: {failures}")
    print(f"总计: {successes + failures}")
    
    return failures == 0

def read_process_output(process, prefix):
    """读取并打印进程的输出"""
    for line in iter(process.stdout.readline, b''):
        print(f"{prefix}: {line.decode().strip()}")

def read_process_error(process, prefix):
    """读取并打印进程的错误输出"""
    for line in iter(process.stderr.readline, b''):
        print(f"{prefix} ERROR: {line.decode().strip()}")

if __name__ == "__main__":
    # 首先启动服务器
    server_process = subprocess.Popen(
        ["python", "-m", "app.main"],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        bufsize=1,
        universal_newlines=True
    )
    
    # 启动线程读取服务器输出
    stdout_thread = threading.Thread(target=read_process_output, args=(server_process, "SERVER"))
    stderr_thread = threading.Thread(target=read_process_error, args=(server_process, "SERVER"))
    stdout_thread.daemon = True
    stderr_thread.daemon = True
    stdout_thread.start()
    stderr_thread.start()
    
    print("等待服务器启动...")
    import time
    time.sleep(3)
    
    try:
        success = test_all_endpoints()
        sys.exit(0 if success else 1)
    finally:
        print("关闭服务器...")
        server_process.terminate()
        time.sleep(1)
        print("服务器关闭")
        server_process.wait(5) 

File: frontend/test-api.js
// 简单的前端API测试脚本
const API_BASE_URL = 'http://localhost:8001/api';

// 模拟fetch请求
async function get(endpoint) {
  try {
    const url = `${API_BASE_URL}${endpoint}`;
    console.log(`发起GET请求: ${url}`);
    
    const response = await fetch(url);
    
    if (!response.ok) {
      throw new Error(`API错误: ${response.status}`);
    }
    
    const data = await response.json();
    console.log(`请求成功: ${endpoint}`);
    console.log(JSON.stringify(data, null, 2));
    return data;
  } catch (error) {
    console.error(`请求失败: ${error}`);
    throw error;
  }
}

async function post(endpoint, data) {
  try {
    const url = `${API_BASE_URL}${endpoint}`;
    console.log(`发起POST请求: ${url}`);
    console.log(`数据: ${JSON.stringify(data)}`);
    
    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data)
    });
    
    if (!response.ok) {
      throw new Error(`API错误: ${response.status}`);
    }
    
    const responseData = await response.json();
    console.log(`请求成功: ${endpoint}`);
    console.log(JSON.stringify(responseData, null, 2));
    return responseData;
  } catch (error) {
    console.error(`请求失败: ${error}`);
    throw error;
  }
}

// 测试所有API端点
async function testAllEndpoints() {
  console.log('===== 开始测试所有API端点 =====');
  
  try {
    // 测试根端点
    console.log('\n----- 测试根端点 -----');
    await get('');
    
    // 测试问题API
    console.log('\n----- 测试问题列表 -----');
    const questions = await get('/questions');
    
    if (questions.items && questions.items.length > 0) {
      const questionId = questions.items[0].id;
      console.log(`\n----- 测试问题详情 (ID: ${questionId}) -----`);
      await get(`/questions/${questionId}`);
    }
    
    // 测试新闻API
    console.log('\n----- 测试新闻列表 -----');
    const news = await get('/news');
    
    if (news.items && news.items.length > 0) {
      const newsId = news.items[0].id;
      console.log(`\n----- 测试新闻详情 (ID: ${newsId}) -----`);
      await get(`/news/${newsId}`);
    }
    
    // 测试学习内容API
    console.log('\n----- 测试学习内容列表 -----');
    await get('/learning');
    
    // 测试聊天API
    console.log('\n----- 测试聊天API -----');
    await post('/chat', { message: '请介绍一下什么是股票?' });
    
    console.log('\n===== 所有API测试完成 =====');
  } catch (error) {
    console.error('\n测试过程中发生错误:', error);
  }
}

async function testApiConnection() {
  try {
    // 测试 API 状态
    const statusResponse = await fetch('http://localhost:8000/api/status');
    const statusData = await statusResponse.json();
    console.log('API Status:', statusData);

    // 测试学习内容 API
    const learningResponse = await fetch('http://localhost:8000/api/learning');
    const learningData = await learningResponse.json();
    console.log('Learning Data:', learningData);

    // 测试新闻 API
    const newsResponse = await fetch('http://localhost:8000/api/news');
    const newsData = await newsResponse.json();
    console.log('News Data:', newsData);

    // 测试问题 API
    const questionsResponse = await fetch('http://localhost:8000/api/questions');
    const questionsData = await questionsResponse.json();
    console.log('Questions Data:', questionsData);

  } catch (error) {
    console.error('API Test Failed:', error);
  }
}

// 执行测试
testAllEndpoints();
testApiConnection(); 